methods/inputs/filter_gedi_data.py

```diff
@@ -9,85 +9,125 @@
 from biomassrecovery.data.gedi_granule import GediGranule  # type: ignore
 from biomassrecovery.constants import WGS84  # type: ignore
 
+# function to parse a single GEDI granule file and return a GeoDataFrame
 def parse_file(file: str) -> gpd.GeoDataFrame:
     with GediGranule(file) as granule:
         granule_data = []
-        for beam in granule.iter_beams():
+        for beam in granule.iter_beams():  # iterate over beams within the granule
             try:
+                # suppress warnings while processing
                 with warnings.catch_warnings():
                     warnings.simplefilter("ignore")
-                    beam.sql_format_arrays()
-                    granule_data.append(beam.main_data)
+                    beam.sql_format_arrays()  # format arrays for the beam data
+                    granule_data.append(beam.main_data)  # collect main data for the beam
             except KeyError:
+                # if a KeyError occurs, skip this beam
                 pass
+        # concatenate beam data into a DataFrame
         df = pd.concat(granule_data, ignore_index=True)
+        # convert to GeoDataFrame using WGS84 coordinates
         gdf = gpd.GeoDataFrame(df, crs=WGS84)
     return gdf
 
+# function to load granule filenames from a text file
+def load_granules_from_file(granule_list_file: str) -> list:
+    """loads the granule filenames from a text file"""
+    with open(granule_list_file, "r") as file:
+        # read each line as a granule name and return the list
+        granule_files = file.read().splitlines()
+    return granule_files
+
+# main function to filter GEDI data based on the buffer and granule list
 def filter_gedi_data(
-    boundary_file: str,
     gedi_directory_path: str,
+    boundary_file: str,
+    granule_list_file: str,
     output_filename: str
 ) -> None:
 
+    # load the boundary file (e.g., shapefile or geojson)
     boundary = gpd.read_file(boundary_file)
 
-    # These are the columns we care about for calculating AGB values, the rest we'll drop
-    # to save memory/disk space
-    interesting_columns = {'sensitivity', 'geometry', 'absolute_time', 'l4_quality_flag', 'degraded_flag',
-        'beam_type', 'agbd'}
+    # specify the columns that are important for AGB calculations
+    interesting_columns = {'sensitivity', 'geometry', 'absolute_time', 'l4_quality_flag', 'degrade_flag',
+                           'beam_type', 'agbd'}
+
+    granules = []  # list to collect data from all granules
+    # load granule names from the provided text file
+    granule_files = load_granules_from_file(granule_list_file)
+
+    # loop through each granule name from the file
+    for granule_name in granule_files:
+        granule_path = os.path.join(gedi_directory_path, granule_name)  # create the full path to the granule
+        if os.path.exists(granule_path):  # check if the granule file exists
```

Comment:
Can you invert this check and just continue if it doesn't exist, saving indenting all the following code?

Reply:
Yes, that would be better, plus returning a specified error message along the lines of "Relevant GEDI Data Missing" would be useful if the granule_path does not exist in the files.

Reply:
```
for granule_name in granule_files:
        granule_path = os.path.join(gedi_directory_path, granule_name)  # create the full path to the granule
        if not os.path.exists(granule_path):  # inverted check: if the file doesn't exist
            raise FileNotFoundError("Revelant GEDI data Missing") # raise error if missing
        df = parse_file(granule_path)  # parse the granule file into a DataFrame

        # drop any columns not in the interesting_columns set
        df.drop(list(set(df.columns) - interesting_columns), axis=1)

        # filter rows based on beam type, quality flag, degradation flag, and date range
        df.drop(df.index[df.beam_type != "full"], inplace=True)
        df.drop(df.index[df.degrade_flag != 0], inplace=True)
        df.drop(df.index[df.l4_quality_flag != 1], inplace=True)

        # filter based on whether the data point falls within the boundary geometry
        df.drop(df.index[~df.within(boundary.geometry.values[0])], inplace=True)
```

Reply:
Here's an idea for a revised loop

Reply:
Doesn't look like this change made it in?